fri/src/proof.rs

```diff
@@ -3,18 +3,170 @@
+
+        // make sure the remaining domain size matches remainder length
+        if domain_size != num_remainder_elements {
+            return Err(ProofSerializationError::InvalidRemainderDomain(
+                num_remainder_elements,
+                domain_size,
+            ));
+        }
+
+        Ok((layer_queries, layer_proofs))
+    }
+
+    /// Returns a vector of remainder values (last FRI layer)
+    pub fn parse_remainder<E: FieldElement>(&self) -> Result<Vec<E>, ProofSerializationError> {
+        let remainder = read_elements_into_vec(&self.remainder).map_err(|err| {
+            ProofSerializationError::RemainderDeserializationError(err.to_string())
+        })?;
+        Ok(remainder)
+    }
+}
+
+// FRI PROOF LAYER
+// ================================================================================================
+
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
-pub struct FriProof {
-    pub layers: Vec<FriProofLayer>,
-    pub rem_values: Vec<u8>,
-    pub partitioned: bool,
+pub struct FriProofLayer {
+    values: Vec<u8>,
+    paths: Vec<u8>,
+}
+
+impl FriProofLayer {
+    /// Creates a new proof layer from the specified query values and the corresponding Merkle
+    /// paths aggregated into a single batch Merkle proof.
+    pub fn new<E: FieldElement, const N: usize>(
+        query_values: Vec<[E; N]>,
+        merkle_proof: BatchMerkleProof,
+    ) -> Self {
+        assert!(!query_values.is_empty(), "query values cannot be empty");
+
+        // TODO: add debug check that values actually hash into the leaf nodes of the batch proof
+
+        // concatenate all query values together into a single vector of bytes
+        let num_queries = query_values.len();
+        let mut values = Vec::with_capacity(num_queries * N * E::ELEMENT_BYTES);
+        for elements in query_values.iter() {
+            values.extend_from_slice(E::elements_as_bytes(elements));
+        }
+
+        // concatenate all internal proof nodes together into a single vector of bytes; we care
+        // about internal nodes only because leaf nodes can be reconstructed from hashes of query
+        // values
+        let paths = merkle_proof.serialize_nodes();
+
+        FriProofLayer { values, paths }
+    }
+
+    /// Decomposes this layer into a combination of query values and corresponding Merkle
+    /// paths (grouped together into a single batch Merkle proof).
+    pub fn parse<H: Hasher, E: FieldElement>(
+        self,
+        domain_size: usize,
+        folding_factor: usize,
+    ) -> Result<(Vec<E>, BatchMerkleProof), SerializationError> {
+        let hash_fn = H::hash_fn();
+
+        // make sure we have enough bytes to read the expected number of queries
+        let num_query_bytes = E::ELEMENT_BYTES * folding_factor;
+        let num_queries = self.values.len() / num_query_bytes;
+        if self.values.len() % num_query_bytes != 0 {
+            let expected_bytes = (num_queries + 1) * num_query_bytes;
```

Comment:
is this the same with `self.values.len() + num_query_bytes`? Can you crosscheck whether this `if` statement is correct?

Reply:
Not quite: the division on line 143 is a truncating division - so, if the bytes in `self.values` can't be evenly divided into query elements it will be one less than whole number of query values. For example:

* Let's say `num_query_bytes` = 64 and `self.values.len()` = 129
* Then, `num_queries` = 2, and `self.values.len() % num_query_bytes` = 1
* This means there is one extra byte, and thus `expected_bytes` will be computed as 3 * 64 = 192